### PR TITLE
design doc: DaemonSet support in scale-out Arktos

### PR DESCRIPTION
**Pull request by h-w-chen**
**What type of PR is this?**
/kind design
/kind documentation

**What this PR does / why we need it**:
It states the supported DaemonSet resource be of syatem tenant only, clarifies TP be its control plane, depicts the necessary changes that should be made to Arktos components.

**Does this PR introduce a user-facing change?**:
NONE (design doc only)

### CODE REVIEW

File: docs/design-proposals/arch/arktos_scale_out.md [hunk 1]
```diff
@@ -122,6 +124,11 @@ A resource manager contains the following components:
 * Active-standby Controller manager process
 * Active-standby scheduler
 
+### <a id="daemonset"></a>DaemonSet in Scale-out Arktos
+In the scale-out architecture, DaemonSet is only allowed for system tenant. A DaemonSet object resides on a Tenant Partition, in the TP's system tenant (by the priciple of no single "system partition", there exist multiple TP that have system tenants); DaemonSet controller and scheduler deployed on the TP create daemon Pods and assign then with proper node resource, respectively. Kubelet looks up the specific TP for API resources (e.g. configmap, secret) needed by the initialization of  system-tenanted daemon Pod.
```
Comment (yb01) [inline docs/design-proposals/arch/arktos_scale_out.md:128]: Even if users deploy Arktos in the traditional single cluster, the DS should be still allowed for system tenant ( or admin ) of the cluster. Scale-out only makes it more complex to manage.

Comment (h-w-chen) [reply docs/design-proposals/arch/arktos_scale_out.md:128]: yes DS in system tenant should be allowed and supported in the so-called scale-up Arktos.
Good point about that clarification; I'll record it in the doc.

File: docs/design-proposals/arch/daemonset.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+## DaemonSet in Arktos
+
+Due to the nature Arktos being a multiple-tenanted system, DaemonSet is only allowed in system tenant; only cluster admin is permitted to create/delete DaemonSet resources.
+
+### DaemonSet in Scale-up Arktos
+Scale-up Arktos has single master. With the admission controller blocking non-system tenant DaemonSet enabled by API Server, Arktos is able to ensure DaemonSet proper support.
+
+### DaemonSet in Scale-out Arktos
+[Scale-out Arktos](arktos_scale_out.md) may have multiple tenant partitions and resource managers; each of them has Kubernetes master components including API Server, controller manager, etc. DaemonSet resource will be managed by tenant partition, as the controllers in tenant partition have the global view of Node resources.
```
Comment (yb01) [inline docs/design-proposals/arch/daemonset.md:9]: probably it makes the naming consistent by using "resource partitions" instead of "resource managers"
also, Arktos should be used instead of Kubernetes in the docs.

File: docs/design-proposals/arch/daemonset.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+In tenant partition, DaemonSet controller populates daemon Pods for the specific DaemonSet, using the global view of Nodes; its scheduler schedules each daemon Pod to the proper Node subsequently.
```
Comment (yb01) [inline docs/design-proposals/arch/daemonset.md:11]: if a tenant partition is dynamically selected for the daemonset, this design should include a way to select it, i.e. an upper layer component or tool used by the admins.

**Review by yb01 — APPROVED** [review]
/lgtm

File: docs/design-proposals/arch/daemonset.md (hunk 3 above)
Comment (h-w-chen) [reply docs/design-proposals/arch/daemonset.md:11]: will change as suggested.

File: docs/design-proposals/arch/daemonset.md (hunk 2 above)
Comment (h-w-chen) [reply docs/design-proposals/arch/daemonset.md]: changed as suggested

File: docs/design-proposals/arch/daemonset.md (hunk 3 above)
Comment (h-w-chen) [reply docs/design-proposals/arch/daemonset.md:11]: such a mechanism is nice to have; seems out of scope of the core of daemonset support. one trivial way to select is via kubectl config context.

File: docs/design-proposals/arch/arktos_scale_out.md (hunk 1 above)
Comment (h-w-chen) [reply docs/design-proposals/arch/arktos_scale_out.md]: doc restructured as separate file which covers scale-up and scale-up both.

File: docs/design-proposals/arch/daemonset.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
+## DaemonSet in Arktos
+
+Due to the nature Arktos being a multiple-tenanted system, DaemonSet is only allowed in system tenant; only cluster admin is permitted to create/delete DaemonSet resources.
+
+### DaemonSet in Scale-up Arktos
+Scale-up Arktos has single master. With the admission controller blocking non-system tenant DaemonSet enabled by API Server, Arktos is able to ensure DaemonSet proper support.
+
+### DaemonSet in Scale-out Arktos
+[Scale-out Arktos](arktos_scale_out.md) may have multiple tenant partitions and resource partitions; each of them has master components including API Server, controller manager, etc. DaemonSet resource will be managed by tenant partition, as the controllers in tenant partition have the global view of Node resources.
```
Comment (Sindica) [inline docs/design-proposals/arch/daemonset.md:9]: I would suggest document the three solutions we have discussed and compare their pros and cons as we did.

File: docs/design-proposals/arch/daemonset.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+Below is the diagram of DaemonSet admission:
+<p align="center"> <img src="images/daemonset_OPD/daemonset-support-DS-creating-DSadding.jpg"> </p>
+
+
+#### Kubelet
+Kubelet shall be able to identify the API resources (configmap, secret etc) from the same tenant partition as the daemon Pod to be started, in environment preparing process.
+
+Below is the diagram of kubelet environment preparing process:
+<p align="center"> <img src="images/daemonset_OPD/daemonset-support-DS-creating-podStarting-envPreparing.jpg"> </p>
```
Comment (Sindica) [inline docs/design-proposals/arch/daemonset.md:30]: typo in daemonset-support-DS-creating-DSadding.jpg

Comment (h-w-chen) [reply docs/design-proposals/arch/daemonset.md:49]: fixed to "admitting"

File: docs/design-proposals/arch/daemonset.md (hunk 4 above)
Comment (h-w-chen) [reply docs/design-proposals/arch/daemonset.md:9]: added design alternatives as suggested

File: docs/design-proposals/arch/daemonset.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+This architecture has some defects:
+* Lack of built-in mechanism to distribute and sync DaemonSet copies to all tenant partitions;
+* Increase of inconsistency of tenanted API resource accesses (e.g. kubelet need to get Pods from RP besides TP's).
+
+**DaemonSet in Hybrid**
+
+This is a variant of the above-mentioned DS in RP: cluster admin manages DaemonSet at tenant partition; system implements a built-in distribution mechanism (DaemonSet distribution controller) to push DeamonSet change to all resource partitions.
+
+This architecture mitigates the DS sync problem, but it brings more incompatibility of scale-out and scale-up.
```
Comment (Sindica) [inline docs/design-proposals/arch/daemonset.md:32]: Could add more detailed cons

File: docs/design-proposals/arch/daemonset.md [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+#### Alternatives
+We considered other options for DaemonSet in scale-out Arktos.
+
+**DaemonSet in Resource Partition**
+
+DaemonSet resource is managed combinely by all resource partitions; tenant partition is not involved. Each resource partition has its own copy of the DaemonSet resource; its DaemonSet controller and scheduler manage the DaemonSet in (local) scope of its Nodes.
+
+This architecture has some defects:
+* Lack of built-in mechanism to distribute and sync DaemonSet copies to all tenant partitions;
+* Increase of inconsistency of tenanted API resource accesses (e.g. kubelet need to get Pods from RP besides TP's).
```
Comment (Sindica) [inline docs/design-proposals/arch/daemonset.md:26]: Additional cons:
Increased system complexity: scheduler in RPs; need unified view of daemonset for admin.

I don't think "kubelet need to get Pods from RP besides TP's" is an issue. We are using similar idea across all components in scale out. KCM/Scheduler/Kubelet need to connect multiple API Servers

**Review by Sindica — COMMENTED** [review]
We need more discussion on this design. I am worried this would affect the design of system pod and makes it much more complicated

File: docs/design-proposals/arch/daemonset.md [hunk 8]
```diff
@@ -0,0 +1,49 @@
+## DaemonSet in Arktos
+
+Due to the nature Arktos being a multiple-tenanted system, DaemonSet is only allowed in system tenant; only cluster admin is permitted to create/delete DaemonSet resources.
```
[the remark this answers is not in the document]
Comment (h-w-chen) [reply docs/design-proposals/arch/daemonset.md:3]: it should be able to update. will correct this statement.

File: docs/design-proposals/arch/daemonset.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+### DaemonSet in Scale-up Arktos
+Scale-up Arktos has single master. With the admission controller blocking non-system tenant DaemonSet enabled by API Server, Arktos is able to ensure DaemonSet proper support.
```
[the remark this answers is not in the document]
Comment (h-w-chen) [reply docs/design-proposals/arch/daemonset.md:9]: will edit as suggested.

File: docs/design-proposals/arch/daemonset.md [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+On the node, kubelet gets notification of daemon Pod's creation, then prepares environment for the Pod, finally starts the pod in the prepared environment. The environment preparation includes fetching the specified API resources (like secret, configmap) as defined in Pod spec and provisioning in form of files/data in node's filesystem. In order to ensure API resources are fecthed from the proper tenant partition, kubelet shall maintain the mapping of daemon Pod to tenant partition origin; whenever the needed resource of the daemon Pod is fetched, kubelet shall lookup the mapping and use that of the proper tenant partition.
+
+In case of the tenant partition that manages a DaemonSet fails, other tenant partitions are available for cluster admin to manage DaemonSets; the DaemonSet previously managed on the failed TP may not be updated, system functionality being degraded. There is possibilities to recover the affected DaemonSet to functioning TP, but it is out of our current design interests.
```
[the remark this answers is not in the document]
Comment (h-w-chen) [reply docs/design-proposals/arch/daemonset.md]: will elaborate as suggested.

File: docs/design-proposals/arch/daemonset.md (hunk 7 above)
Comment (h-w-chen) [reply docs/design-proposals/arch/daemonset.md:37]: will edit with this block.

File: docs/design-proposals/arch/daemonset.md (hunk 6 above)
Comment (h-w-chen) [reply docs/design-proposals/arch/daemonset.md]: added explanation of discrepancy scale-up does not need such daemon set sync in place.

File: docs/design-proposals/arch/daemonset.md [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+### What is not covered yet
+At this stage of Arktos, PVC/PV for storage volume is not fully supported - out of scope of this DaemonSet design.
+
+### DaemonSet in Scale-up Arktos
+Scale-up Arktos has single master. With the admission controller blocking non-system tenant DaemonSet, Arktos is able to ensure DaemonSet proper support.
+
+### DaemonSet in Scale-out Arktos
+[Scale-out Arktos](arktos_scale_out.md) may have multiple tenant partitions and resource partitions; each of them has master components including API Server, controller manager, etc. DaemonSet resource will be managed by tenant partition, as the controllers in tenant partition have the global view of Node resources.
+
+Usually DaemonSet reuires other API resources to provision its daemon Pods with specific environmental data. For instance, configmap resoure is used to initialize as app config file, secret as app sensitive data. These supporting resources shall be at the same Tenant Partion as the DaemonSet.
```
Comment (Sindica) [inline docs/design-proposals/arch/daemonset.md:14]: typo: reuires

Comment (Sindica) [reply docs/design-proposals/arch/daemonset.md:14]: resoure -> resource

File: docs/design-proposals/arch/daemonset.md [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+Different Daemonsets can reside on different Tenant Partitions. At one TP, DaemonSet should be uniquely named; however, a DaemonSet of TP could have the same name as one managed by aother TP - though this practice is not recommended.
+
+At the specific tenant partition, DaemonSet controller populates daemon Pods for the specific DaemonSet, using the global view of Nodes; its scheduler schedules each daemon Pod to the proper Node subsequently.
+
+On the node, kubelet gets notification of daemon Pod's creation, then prepares environment for the Pod, finally starts the pod in the prepared environment. The environment preparation includes fetching the specified API resources (like secret, configmap) as defined in Pod spec and provisioning in form of files/data in node's filesystem. In order to ensure API resources are fecthed from the proper tenant partition, kubelet shall maintain the mapping of daemon Pod to tenant partition origin; whenever the needed resource of the daemon Pod is fetched, kubelet shall lookup the mapping and use that of the proper tenant partition.
```
Comment (Sindica) [inline docs/design-proposals/arch/daemonset.md:20]: Isn't Daemon Pod origin same as regular pod origin? Can it use existing pod origin management mechanism?

File: docs/design-proposals/arch/daemonset.md [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+Each node may typically run daemon Pods of all DaemonSet from all the TP's. These daemon Pods should not conflict with each other to compete for same system resources, e.g. no two are trying to set up network as CNI plugin; whether their DaemonSet names are same or not is immaterial.
+
+In case of the tenant partition that manages a DaemonSet fails, other tenant partitions, if applicable, are available for cluster admin to manage DaemonSets; the DaemonSet previously managed on the failed TP may not be updated, system functionality being degraded. There is possibilities to recover the affected DaemonSet to functioning TP, but it is out of our current design interests.
+
+Arranging Daemonset and its supporting resources within the unit of TP as depeicted above does not put unnecessary constraint of how system tenant being managed across TP. Whether Arktos scale-out chooses to have multiple TP's with each being able to manage its own system tenant resources, or to have a single dedicated Partition for system tenant resources likely enhanced by some sort of HA, this design stands - as long as DaemonSet and its supporting API resources are managed by the same TP.
```
Comment (Sindica) [inline docs/design-proposals/arch/daemonset.md:26]: across TP -> across TPs, multiple TP's -> multiple TPs

File: docs/design-proposals/arch/daemonset.md [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+
+Providing a centralized DaemonSet management tool on top of all and every TP is convenience gain for cluster admin; however it is beyond the core design. The rudimentary utility using kubectl config context is able to mitigate before a better solution is available.
+
+#### Alternatives
```
Comment (Sindica) [inline docs/design-proposals/arch/daemonset.md:30]: Alternatives can use a bigger font

File: docs/design-proposals/arch/daemonset.md [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+We considered other options for DaemonSet in scale-out Arktos.
+
+**DaemonSet in Resource Partition**
+
+DaemonSet resource is managed combinely by all resource partitions; tenant partition is not involved. Each resource partition has its own copy of the DaemonSet resource; its DaemonSet controller and scheduler manage the DaemonSet in (local) scope of its Nodes.
+
+This architecture has some defects:
+* Lack of built-in mechanism to distribute and sync DaemonSet copies to all tenant partitions;
+* Increased operation complexity: multiple RP's each runs master controllers;
+* Decreased scheudule accuracy: scheduler in RP schedules daemon Pods which cannot have the resource allocation reflected in schedule in TP, which leads to inaccuracy of the deceived resources.
```
Comment (Sindica) [inline docs/design-proposals/arch/daemonset.md:40]: typo: scheudule

File: docs/design-proposals/arch/daemonset.md [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+* Overlap of TP/RP: Pod/configmap/secret may be managed by TP or RP - depending on whether they are DaemonSet related or not, which tarnishes the clear cut of TP/RP.
+
+**DaemonSet in Hybrid**
+
+This is a variant of the above-mentioned DS in RP: cluster admin manages DaemonSet at tenant partition; system implements a built-in distribution mechanism (DaemonSet distribution controller) to push DeamonSet change to all resource partitions.
+
+This architecture mitigates the DS sync problem, but it brings more incompatibility of scale-out and scale-up, as scale-up hs no such sync in the first place.
```
Comment (Sindica) [inline docs/design-proposals/arch/daemonset.md:47]: typo: hs

File: docs/design-proposals/arch/daemonset.md [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+
+
+#### Kubelet
+Kubelet shall be able to identify the API resources (configmap, secret, PVC etc) from the same tenant partition as the daemon Pod to be started, in environment preparing process.
+
+Below is the diagram of kubelet environment preparing process:
+<p align="center"> <img src="images/daemonset_OPD/daemonset-support-DS-creating-podStarting-envPreparing.jpg"> </p>
+
+There are two significances for kubelet, at high level:
+1. Keeping track of daemon Pod's TP origin.
```
Comment (Sindica) [inline docs/design-proposals/arch/daemonset.md:67]: Isn't it the same as regular tenant pod?

File: docs/design-proposals/arch/daemonset.md [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+<p align="center"> <img src="images/daemonset_OPD/daemonset-support-DS-creating-DSadding.jpg"> </p>
+
+
+#### Kubelet
+Kubelet shall be able to identify the API resources (configmap, secret, PVC etc) from the same tenant partition as the daemon Pod to be started, in environment preparing process.
+
+Below is the diagram of kubelet environment preparing process:
+<p align="center"> <img src="images/daemonset_OPD/daemonset-support-DS-creating-podStarting-envPreparing.jpg"> </p>
+
+There are two significances for kubelet, at high level:
```
Comment (Sindica) [inline docs/design-proposals/arch/daemonset.md:66]: What do you meant by "significances"?

**Review by Sindica — APPROVED** [review]
Generally it looks good. Minor comments below.

File: docs/design-proposals/arch/daemonset.md (hunk 11 above)
Comment (h-w-chen) [reply docs/design-proposals/arch/daemonset.md:14]: will fix them.

File: docs/design-proposals/arch/daemonset.md (hunk 12 above)
Comment (h-w-chen) [reply docs/design-proposals/arch/daemonset.md:20]: will add a section to clarify this.

File: docs/design-proposals/arch/daemonset.md (hunk 13 above)
Comment (h-w-chen) [reply docs/design-proposals/arch/daemonset.md:26]: will fix

File: docs/design-proposals/arch/daemonset.md (hunk 14 above)
Comment (h-w-chen) [reply docs/design-proposals/arch/daemonset.md:30]: will change as suggested

File: docs/design-proposals/arch/daemonset.md (hunk 15 above)
Comment (h-w-chen) [reply docs/design-proposals/arch/daemonset.md:40]: will fix

File: docs/design-proposals/arch/daemonset.md (hunk 16 above)
Comment (h-w-chen) [reply docs/design-proposals/arch/daemonset.md:47]: has

File: docs/design-proposals/arch/daemonset.md (hunk 17 above)
Comment (h-w-chen) [reply docs/design-proposals/arch/daemonset.md:67]: will add a sentence to clarify.

File: docs/design-proposals/arch/daemonset.md (hunk 18 above)
Comment (h-w-chen) [reply docs/design-proposals/arch/daemonset.md:66]: will reword it to "significant required functionalities"
